graph/src/blockchain/mod.rs

```diff
@@ -255,15 +255,31 @@ pub trait DataSource<C: Blockchain>: 'static + Sized + Send + Sync + Clone {
 
     fn address(&self) -> Option<&[u8]>;
     fn start_block(&self) -> BlockNumber;
+    fn end_block(&self) -> Option<BlockNumber>;
     fn name(&self) -> &str;
     fn kind(&self) -> &str;
     fn network(&self) -> Option<&str>;
     fn context(&self) -> Arc<Option<DataSourceContext>>;
     fn creation_block(&self) -> Option<BlockNumber>;
     fn api_version(&self) -> semver::Version;
+
+    /// **NOTE: DO NOT OVERRIDE THIS METHOD.**
+    /// Use `min_spec_version_inner` for custom logic instead.
+    /// This method is used to determine the minimum spec version required by the data source.
     fn min_spec_version(&self) -> semver::Version {
```

Comment:
This does not look like the right place for this sort of contraint, maybe the manifest validation would be the right place. This here is leaking the details about spec into data sources

Reply:
Got it, i moved this logic to manifest validation